### PR TITLE
complete the implementation of cleanUTF8

### PR DESCRIPTION
**Pull request by cgao-slack**
###  Summary

Complete the implementation of cleanUTF8 since there are few special characters in our blog posts.

Lines that are changed from php html-sanitizer:
https://github.com/ezyang/htmlpurifier/blob/master/library/HTMLPurifier/Encoder.php#L160
https://github.com/ezyang/htmlpurifier/blob/master/library/HTMLPurifier/Encoder.php#L162
https://github.com/ezyang/htmlpurifier/blob/master/library/HTMLPurifier/Encoder.php#L236

### Requirements (place an `x` in each `[ ]`)

* [x] I've read and understood the [Contributing Guidelines](https://github.com/slackhq/htmlsanitizer-hack/blob/main/.github/contributing.md) and have done my best effort to follow them.
* [x] I've read and agree to the [Code of Conduct](https://slackhq.github.io/code-of-conduct).
* [ ] I've written tests to cover the new code and functionality included in this PR.

### CODE REVIEW

File: src/Encoder.hack [hunk 1]
```diff
@@ -117,9 +117,149 @@ class HTMLPurifier_Encoder {
             return $str;
         }
 
-        // If anything does not match the regex, it is the 1% that doesn't and thus we really do NOT 
-        // want to be supporting that, so we just throw an exception instead
-        throw new \Exception('This string is in the 1% that does not fall in the main case');
+		$mState = 0; // cached expected number of octets after the current octet
```
Comment (og-slack) [inline src/Encoder.hack:120]: Is the indentation funky here or is that just GitHub? Looks like tabs vs spaces?

**Comment by og-slack** [comment]
Is it possible to add some short tests for this new functionality?

File: src/Encoder.hack (hunk 1 above)
Comment (cgao-slack) [reply src/Encoder.hack]: Yes, there is an indentation problem.

File: tests/HTMLPurifierTest.hack [hunk 2]
```diff
@@ -103,12 +103,19 @@ class HTMLPurifierTest extends HackTest {
 </table>';
 		$purifier = new HTMLPurifier\HTMLPurifier($config);
 		$clean_html = $purifier->purify($dirty_html);
-		expect($clean_html)->toEqual('<table><caption>
+		expect($clean_html)->toEqual('<table>
+  <caption>
     Cool table
   </caption>
-  <tfoot><tr><th>I can do so much!</th>
```
Comment (og-slack) [inline tests/HTMLPurifierTest.hack:109]: I think changing this breaks the test, right? :laughcry:

File: tests/HTMLPurifierTest.hack [hunk 3]
```diff
@@ -320,4 +327,55 @@ class HTMLPurifierTest extends HackTest {
         expect(true)->toNotBeNull();
 		echo "finished.\n\n";
     }
+
+	public function testWebappInput(): void {
```
Comment (og-slack) [inline tests/HTMLPurifierTest.hack:331]: Let's use a more generic name for this

File: tests/HTMLPurifierTest.hack (hunk 2 above)
Comment (cgao-slack) [reply tests/HTMLPurifierTest.hack:109]: Hmm, very strange that test failed. When I run it using my local hhvm, they passed, but in container version of hhvm it failed.

Comment (og-slack) [reply tests/HTMLPurifierTest.hack:109]: Yeah, this is the issue we've been talking about in the Hacklang workspace.  Something strange happens with the spacing between versions or between Mac/Linux

**Review by og-slack — APPROVED** [review]

**Pull request merged**